[JENKINS-58779] LogRotator stops, if build was deleted externally

HedAurabesh opened the pull request:
The deletion of a build/run in Jenkins is handled by the delete() method.

It throws an exception, when the build directory on disk has already been
removed and stops deleting the build.

Unfortunately, the hudson.tasks.LogRotator.perform(Job<?, ?>) method does
not check for this exception and will STOP cleaning up, once even just one
build throws that exception.

This behaviour is fixed in the following ways in this change:

1.) The Run.delete() method must still unlink the job, even if the build
    directory has been removed externally.

2.) The log output of Run.delete() must be made less verbose (it prints the
    ENTIRE folder content by default, into the job that is doing the cleanup)

3.) The LogRotator.perform() method must handle the exceptions being thrown
    by delete() gracefully.

Signed-off-by: Martin H Schroeder <martin.h.schroeder@intel.com>

See [JENKINS-58779](https://issues.jenkins-ci.org/browse/JENKINS-58779).

HedAurabesh committed:
[JENKINS-58779] LogRotator stops cleaning, if build was deleted externally

The deletion of a build/run in Jenkins is handled by the delete() method.

It throws an exception, when the build directory on disk has already been
removed and stops deleting the build.

Unfortunately, the hudson.tasks.LogRotator.perform(Job<?, ?>) method does
not check for this exception and will STOP cleaning up, once even just one
build throws that exception.

This behaviour is fixed in the following ways in this change:

1.) The Run.delete() method must still unlink the job, even if the build
    directory has been removed externally.

2.) The log output of Run.delete() must be made less verbose (it prints the
    ENTIRE folder content by default, into the job that is doing the cleanup)

3.) The LogRotator.perform() method must handle the exceptions being thrown
    by delete() gracefully.

Signed-off-by: Martin H Schroeder <martin.h.schroeder@intel.com>

HedAurabesh committed:
Small change to copyright header (removed blank line)

n/t

Signed-off-by: Martin H Schroeder <martin.h.schroeder@intel.com>

oleg-nenashev commented on core/src/main/java/hudson/model/Run.java:
```diff
@@ -1566,34 +1573,52 @@ public synchronized void deleteArtifacts() throws IOException {
+            }
+            boolean renamingSucceeded = true;
+            try {
+                Path path = Files.move(rootDir.toPath(), tmp.toPath(), StandardCopyOption.ATOMIC_MOVE);
```
`InvalidPathException` is not handled here. I recommend using `Util#fileToPath()` instead. https://github.com/jenkinsci/jenkins/blob/bfdd49537cf84f94a17559f2bd6c105d2b5e7c3e/core/src/main/java/hudson/Util.java#L1499

```suggestion
                Path path = Files.move(rootDir.toPath(), Util.fileToPath(), StandardCopyOption.ATOMIC_MOVE);
```

HedAurabesh replied:
Done

oleg-nenashev commented on core/src/main/java/hudson/model/Run.java:
```diff
@@ -1566,34 +1573,52 @@ public synchronized void deleteArtifacts() throws IOException {
+            try {
+                Path path = Files.move(rootDir.toPath(), tmp.toPath(), StandardCopyOption.ATOMIC_MOVE);
+            } catch (UnsupportedOperationException | IOException | SecurityException ex) {
+                // Fall back to previous < Java 7 variant
```
Add FINE logging here?

HedAurabesh replied:
Done

oleg-nenashev commented on core/src/main/java/hudson/tasks/LogRotator.java:
```diff
@@ -108,8 +109,44 @@ public LogRotator(int daysToKeep, int numToKeep, int artifactDaysToKeep, int art
+     * This method calls {@link Run#delete()} with all checked exceptions
+     * properly caught and reported in the log.
+     * 
+     * @param r the run to delete, must not be null.
```
```suggestion
     * @param r the run to delete, must not be null.
     * @since TODO
```

Protected API of a public class is a public API

HedAurabesh replied:
Done

oleg-nenashev commented on core/src/main/java/hudson/tasks/LogRotator.java:
```diff
@@ -108,8 +109,44 @@ public LogRotator(int daysToKeep, int numToKeep, int artifactDaysToKeep, int art
+     * This method calls {@link Run#deleteArtifacts()} with all checked exceptions
+     * properly caught and reported in the log.
+     * 
+     * @param r the run for which to delete artifacts, must not be null.
```
```suggestion
     * @param r the run for which to delete artifacts, must not be null.
     * @since TODO
```

HedAurabesh replied:
Done

oleg-nenashev reviewed (changes requested):
Just one minor comment about error handling..

I also think that `perform()` should throw an exception when such error happens. I suggest to accumulate exceptions in the method and then to throw an aggregated `IOExceptions` for all builds/artifacts the rotator was unable to delete. It can be achieved by creating a new Exception object and adding suppressed exceptions there

HedAurabesh commented:
Hi Oleg. Thanks for the review. I will incorporate the changes and push a new commit as soon as I am able to do so (I am currently travelling).

HedAurabesh committed:
Incorporated changes proposed by Oleg Nenashev

 - Usage of Util.fileToPath
 - Addition of @since headers to public API methods

Signed-off-by: Martin H Schroeder <martin.h.schroeder@intel.com>

HedAurabesh commented:
@oleg-nenashev : The changes proposed by you have been commited and uploaded.

HedAurabesh commented:
> Just one minor comment about error handling..
> 
> I also think that `perform()` should throw an exception when such error happens. I suggest to accumulate exceptions in the method and then to throw an aggregated `IOExceptions` for all builds/artifacts the rotator was unable to delete. It can be achieved by creating a new Exception object and adding suppressed exceptions there

Hi Oleg, I have also prepared (but not uploaded) a commit for that change. The issue I have with that is mainly, that LogRotator might be overidden by subclasses and changing the method signature by adding a new checked exception is not nice.

We could use an unchecked exception, instead. That is not nice either, but at least does not break compatibility.

jeffret-b commented on core/src/main/java/hudson/model/Run.java:
```diff
@@ -1566,34 +1573,62 @@ public synchronized void deleteArtifacts() throws IOException {
+                        StandardCopyOption.ATOMIC_MOVE
+                );
+            } catch (UnsupportedOperationException | IOException | SecurityException ex) {
+                // Fall back to previous < Java 7 variant
```
There really isn't any reason to support anything earlier than Java 8. I would use the 7+ variant and not bother complicating the code with supporting anything earlier.

HedAurabesh replied:
Done

jeffret-b reviewed (approved):
Looks good. I haven't tested it out but it looks like it should do the job.

jeffret-b commented:
> Hi Oleg, I have also prepared (but not uploaded) a commit for that change. The issue I have with that is mainly, that LogRotator might be overidden by subclasses and changing the method signature by adding a new checked exception is not nice.
> 
> We could use an unchecked exception, instead. That is not nice either, but at least does not break compatibility.

@HedAurabesh , I'm confused by your comment here. As I read the diff, you removed the declared IOException in this PR. All you need to do is restore that declaration and then create a new, aggregated exception at the end after the `perform()` has done all it can.

HedAurabesh commented:
> @HedAurabesh , I'm confused by your comment here. As I read the diff, you removed the declared IOException in this PR. All you need to do is restore that declaration and then create a new, aggregated exception at the end after the `perform()` has done all it can.
Hi Jeff. You are indeed absolutely right. It's been a while since I wrote the initial change, so I totally forgot that the IOException was removed. 🤦‍♂ 

I am going to slightly alter my change and soon upload the change that has the "combined reporting IOException".

HedAurabesh committed:
Added fixes for comments raised by Oleg Nenashev and Jeff Thompson

This adds a collective IOException raised for all sub-exceptions
encountered during the cleanup process.

Signed-off-by: Martin H Schroeder <martin.h.schroeder@intel.com>

HedAurabesh commented:
@oleg-nenashev, @jeffret-b : I have uploaded a commit with your requested changes implemented.

jeffret-b reviewed (approved):
The earlier items were cleaned up nicely. This should work well.

I read the code but, haven't tested it any further.

HedAurabesh commented:
@oleg-nenashev: Can you resolve your review comments? Thanks!

oleg-nenashev commented on core/src/main/java/hudson/model/Run.java:
```diff
@@ -1566,34 +1569,51 @@ public synchronized void deleteArtifacts() throws IOException {
-            throw new IOException(this + ": " + rootDir + " looks to have already been deleted; siblings: " + Arrays.toString(project.getBuildDir().list()));
+            //No root directory found to delete. Somebody seems to have nuked
+            //it externally. Logging a warning before dropping the build
+            LOGGER.warning(String.format(
```
This pattern requires the string formatting even if the message is not printed. `LOGGER.log(Level.WARNING, <pattern>, new Obcject[] {args})` is a better pattern. Not an issue for WARNINGs tho, they are printed in all common cases

oleg-nenashev commented on core/src/main/java/hudson/model/Run.java:
```diff
@@ -1566,34 +1569,51 @@ public synchronized void deleteArtifacts() throws IOException {
+            // some user reported that they see some left-over .xyz files in the workspace,
+            // so just to make sure we've really deleted it, schedule the deletion on VM exit, too.
+            if (tmp.exists()) {
+                tmp.deleteOnExit();
```
I am not against it. but it is rather a separate change

oleg-nenashev reviewed (approved):
It should do the job. Will merge tomorrow if no negative feedback

oleg-nenashev commented:
Thanks a lot @HedAurabesh ! It is great to finally merge it

daniel-beck commented:
No changelog entry 🙁
